src/core/lib/security/credentials/tls/grpc_tls_crl_provider.cc

```diff
@@ -114,5 +222,110 @@ std::shared_ptr<Crl> StaticCrlProvider::GetCrl(
+    std::function<void(absl::Status)> reload_error_callback) {
+  // TODO(gtcooke94) - testing with std::chrono
+  if (refresh_duration < std::chrono::seconds(1)) {
+    return absl::InvalidArgumentError("Refresh duration minimum is 60 seconds");
+  }
+  struct stat dir_stat;
+  if (stat(directory.data(), &dir_stat) != 0) {
+    return absl::InvalidArgumentError("The directory path is not valid.");
+  }
+  auto provider = std::make_shared<DirectoryReloaderCrlProviderImpl>(
+      directory, refresh_duration, reload_error_callback);
+  // TODO(gtcooke94) this could be slow to do at startup, but I think we want to
+  // make sure it's done before the provider is used.
+  absl::Status initial_status = provider->Update();
+  if (!initial_status.ok()) {
+    return initial_status;
+  }
+  provider->ScheduleReload();
+  return provider;
+}
+
+void DirectoryReloaderCrlProviderImpl::OnNextUpdateTimer() {
+  absl::Status status = Update();
+  if (!status.ok()) {
+    if (reload_error_callback_ != nullptr) {
+      reload_error_callback_(status);
+    }
+  }
+  ScheduleReload();
+}
+
+void DirectoryReloaderCrlProviderImpl::ScheduleReload() {
+  std::weak_ptr<DirectoryReloaderCrlProviderImpl> self = shared_from_this();
+  refresh_handle_ = event_engine_->RunAfter(refresh_duration_, [self]() {
+    ApplicationCallbackExecCtx callback_exec_ctx;
+    ExecCtx exec_ctx;
+    {
+      if (std::shared_ptr<DirectoryReloaderCrlProviderImpl> valid_ptr =
+              self.lock()) {
+        valid_ptr->OnNextUpdateTimer();
+      }
+    }
+  });
+}
+
+absl::Status DirectoryReloaderCrlProviderImpl::Update() {
+  std::vector<FileData> crl_files = GetFilesInDirectory(crl_directory_);
+  bool all_files_successful = true;
+
+  absl::flat_hash_map<std::string, std::shared_ptr<Crl>> new_crls;
+  for (const FileData& file : crl_files) {
+    // Build a map of new_crls to update to. If all files successful, do a
+    // full swap of the map. Otherwise update in place
+    absl::StatusOr<std::shared_ptr<Crl>> result = ReadCrlFromFile(file.path);
+    if (!result.ok()) {
+      all_files_successful = false;
+      if (reload_error_callback_ != nullptr) {
+        reload_error_callback_(absl::InvalidArgumentError(absl::StrFormat(
+            "CRL Reloader failed to read file: %s", file.path)));
+      }
+      continue;
+    }
+    // Now we have a good CRL to update in our map
+    std::shared_ptr<Crl> crl = *result;
+    new_crls[crl->Issuer()] = std::move(crl);
+  }
+  if (!all_files_successful) {
+    // Need to make sure CRLs we read successfully into new_crls are still
+    // in-place updated in crls_
+    for (auto& kv : new_crls) {
+      std::shared_ptr<Crl> crl = kv.second;
+      mu_.Lock();
+      crls_[crl->Issuer()] = std::move(crl);
+      mu_.Unlock();
+    }
+    return absl::UnknownError(
+        "Not all files in CRL directory read successfully during async "
+        "update.");
+  } else {
+    mu_.Lock();
```

Comment:
nit: Prefer a mutex lock here instead, so e.g. this could be
```
absl::MutexLock lock(&mu_);
crls_ = std::move(new_crls);
```

Reply:
Changed to have a MutexLock at a higher scope